dev/env.go

```diff
@@ -73,51 +115,54 @@ func (env *Environment) Init(ctx context.Context) error {
 		return err
 	}
 
-	kindConfig := `kind: Cluster
-apiVersion: kind.x-k8s.io/v1alpha4
-`
-
-	// Workaround for https://github.com/kubernetes-sigs/kind/issues/2411
-	// For BTRFS on LUKS.
-	if _, err := os.Lstat("/dev/dm-0"); err == nil {
-		kindConfig += `nodes:
-- role: control-plane
-  extraMounts:
-    - hostPath: /dev/dm-0
-      containerPath: /dev/dm-0
-      propagation: HostToContainer
-`
+	if env.config.KindClusterConfig == nil {
+		return fmt.Errorf("no KinD cluster config found")
 	}
 
 	if err := os.MkdirAll(env.WorkDir, os.ModePerm); err != nil {
 		return fmt.Errorf("creating workdir: %w", err)
 	}
 
+	kindConfigYamlBytes, err := yaml.Marshal(env.config.KindClusterConfig)
+	if err != nil {
+		return fmt.Errorf("failed to process the KinD cluster config as a YAML: %w", err)
+	}
+
 	kubeconfigPath := path.Join(env.WorkDir, "kubeconfig.yaml")
 	kindconfigPath := path.Join(env.WorkDir, "/kind.yaml")
 	if err := ioutil.WriteFile(
-		kindconfigPath, []byte(kindConfig), os.ModePerm); err != nil {
+		kindconfigPath, kindConfigYamlBytes, os.ModePerm); err != nil {
 		return fmt.Errorf("creating kind cluster config: %w", err)
 	}
 
-	// Needs cluster creation?
-	var checkOutput bytes.Buffer
-	if err := env.execKindCommand(ctx, &checkOutput, nil, "get", "clusters"); err != nil {
-		return fmt.Errorf("getting existing kind clusters: %w", err)
+	provider, err := env.getKindProvider()
+	if err != nil {
+		return err
+	}
+
+	// check if cluster already exists with the same name
+	createCluster := true
+	existingKindClusters, err := provider.List()
+	if err != nil {
+		return fmt.Errorf("failed to fetch the existing KinD clusters: %w", err)
+	}
+	for _, cluster := range existingKindClusters {
+		if env.Name == cluster {
+			createCluster = false
+			break
+		}
```

Comment:
@yashvardhan-kukreja just wondering, should this throw an error, if the cluster already exists, or does it work either way?

Reply:
Good question Petr. 

This method is not only intended for the "creation" of the KinD cluster/environment but to also serialize it into a devkube-compliant object of the `type Environment` which the client can use further with the countless other functionalities of devkube. So, this gives the client the power to play with devkube on an existing cluster as well. If we throw an error here whenever devkube sees an existing cluster with the same name, this would be mean that we're enforcing the fact that "devkube can only be used on the clusters which are created by devkube". This loses on the user experience because as a client/customer of devkube, I do not want to always create a fully-fledged KinD cluster from scratch; sometimes I just don't devkube to adopt one of my existing (already created) KinD clusters and use its methods on top of it.

Reply:
Makes sense. Thank you for the explanation!